File: nodns-resolver/src/parse.ts
import { npubEncode } from 'nostr-tools/nip19';
import type { NostrEvent, NostrDnsRecord, RecordInfo, SpecVersion, ValidityInfo } from './types.js';
import { DEFAULT_ZONE, VALID_RECORD_TYPES } from './types.js';

function safeNpubEncode(pubkey: string): string {
  try {
    return npubEncode(pubkey);
  } catch {
    return pubkey.slice(0, 16);
  }
}

export function isNpubDerivedName(name: string): boolean {
  return name === '' || name === '@';
}

export function computeFqdn(name: string, pubkey: string, zone: string = DEFAULT_ZONE): string {
  if (isNpubDerivedName(name)) {
    return `${safeNpubEncode(pubkey)}.${zone}`;
  }
  return `${name}.${zone}`;
}

function parseTtl(tag: string[]): number {
  if (tag.length > 10) {
    const parsed = parseInt(tag[10], 10);
    if (!Number.isNaN(parsed) && parsed > 0) return parsed;
  }
  if (tag.length > 4) {
    for (let i = tag.length - 1; i >= 4; i--) {
      const parsed = parseInt(tag[i], 10);
      if (!Number.isNaN(parsed) && parsed > 0) return parsed;
    }
  }
  return 3600;
}

export function parseRecordsFromEvent(
  event: NostrEvent,
  zone: string,
): NostrDnsRecord[] {
  const records: NostrDnsRecord[] = [];

  for (const tag of event.tags) {
    if (tag[0] !== 'record') continue;
    if (tag.length < 4) continue;

    const type = tag[1];
    const name = tag[2] || '';
    const value = tag[3];
    const ttl = tag.length >= 11 ? parseInt(tag[10], 10) : 3600;

    if (!type || !value) continue;

    const fqdn = computeFqdn(name, event.pubkey, zone);

    records.push({
      type,
      name: name || '@',
      value,
      ttl: Number.isNaN(ttl) ? 3600 : ttl,
      fqdn,
      pubkey: event.pubkey,
      eventId: event.id,
      created_at: event.created_at,
    });
  }

  return records;
}

export function parseRecords(
  event: NostrEvent,
  zone: string = DEFAULT_ZONE,
): RecordInfo[] {
  const records: RecordInfo[] = [];

  for (const tag of event.tags) {
    if (tag[0] !== 'record') continue;
    if (tag.length < 4) continue;

    const type = (tag[1] ?? '').toUpperCase();
    const name = tag[2] ?? '';
    const rdata = tag[3] ?? '';
    const ttl = parseTtl(tag);
    const isNpubDerived = isNpubDerivedName(name);
    const fqdn = computeFqdn(name, event.pubkey, zone);

    records.push({ type, name, ttl, rdata, fqdn, isNpubDerived });
  }

  return records;
}

export function parseRecord(
  event: NostrEvent,
  zone: string = DEFAULT_ZONE,
): RecordInfo | null {
  const records = parseRecords(event, zone);
  return records.length > 0 ? records[0] : null;
}

function detectSpecVersion(tags: string[][]): SpecVersion {
  const hasAlt = tags.some((t) => t[0] === 'alt');
  const cashuTag = tags.find((t) => t[0] === 'cashu');
  const hasP2PK = cashuTag?.[1]?.includes('P2PK') ?? false;
  if (hasP2PK) return 'v2';
  if (hasAlt) return 'v1.1';
  return 'v1';
}

export function checkValidity(event: NostrEvent): ValidityInfo {
  const tags = event.tags;
  const specVersion = detectSpecVersion(tags);

  const recordTags = tags.filter((t) => t[0] === 'record');

  if (recordTags.length === 0) {
    return { valid: false, reason: 'no record tags', specVersion };
  }

  for (const tag of recordTags) {
    if (tag.length < 4) {
      return { valid: false, reason: `malformed (${tag.length} fields)`, specVersion };
    }
    const type = (tag[1] ?? '').toUpperCase();
    if (!VALID_RECORD_TYPES.includes(type)) {
      return { valid: false, reason: `unknown type: ${type}`, specVersion };
    }
  }

  return { valid: true, specVersion };
}

export function deduplicateRecords(records: NostrDnsRecord[]): NostrDnsRecord[] {
  const seen = new Map<string, NostrDnsRecord>();
  for (const r of records) {
    const key = `${r.fqdn}:${r.type}:${r.name}:${r.value}`;
    const existing = seen.get(key);
    if (!existing || r.created_at > existing.created_at) {
      seen.set(key, r);
    }
  }
  return Array.from(seen.values()).sort((a, b) => b.created_at - a.created_at);
}

File: nodns-resolver-rs/src/parse.rs
//! Nostr event parsing — record extraction and FQDN computation.

use nostr_sdk::{Event, PublicKey, ToBech32};

use crate::types::{NostrDnsRecord, DEFAULT_ZONE};

const DEFAULT_TTL: u32 = 3600;

#[must_use]
pub fn is_npub_derived_name(name: &str) -> bool {
    name.is_empty() || name == "@"
}

fn safe_npub_encode(pubkey_hex: &str) -> String {
    PublicKey::from_hex(pubkey_hex)
        .ok()
        .and_then(|pk| pk.to_bech32().ok())
        .unwrap_or_else(|| pubkey_hex.chars().take(16).collect())
}

#[must_use]
pub fn compute_fqdn(name: &str, pubkey: &str, zone: &str) -> String {
    if is_npub_derived_name(name) {
        format!("{}.{}", safe_npub_encode(pubkey), zone)
    } else {
        format!("{}.{}", name, zone)
    }
}

#[must_use]
pub fn compute_fqdn_default(name: &str, pubkey: &str) -> String {
    compute_fqdn(name, pubkey, DEFAULT_ZONE)
}

fn parse_ttl_from_tag(tag: &[String]) -> u32 {
    if tag.len() > 10 {
        if let Ok(parsed) = tag[10].parse::<u32>() {
            if parsed > 0 {
                return parsed;
            }
        }
    }
    if tag.len() > 4 {
        for i in (4..tag.len()).rev() {
            if let Ok(parsed) = tag[i].parse::<u32>() {
                if parsed > 0 {
                    return parsed;
                }
            }
        }
    }
    DEFAULT_TTL
}

#[must_use]
pub fn parse_records_from_event(event: &Event, zone: &str) -> Vec<NostrDnsRecord> {
    let mut records = Vec::new();
    let pubkey_hex = event.pubkey.to_hex();
    let event_id = event.id.to_hex();
    let created_at = event.created_at.as_secs() as i64;

    for tag in event.tags.iter() {
        let slice = tag.as_slice();
        if slice.is_empty() || slice[0] != "record" {
            continue;
        }
        if slice.len() < 4 {
            continue;
        }

        let record_type = slice[1].clone();
        let name = if slice[2].is_empty() {
            "@".to_string()
        } else {
            slice[2].clone()
        };
        let value = slice[3].clone();

        if record_type.is_empty() || value.is_empty() {
            continue;
        }

        let ttl = parse_ttl_from_tag(slice);
        let fqdn = compute_fqdn(&name, &pubkey_hex, zone);

        records.push(NostrDnsRecord {
            record_type,
            name,
            value,
            ttl,
            fqdn,
            pubkey: pubkey_hex.clone(),
            event_id: event_id.clone(),
            created_at,
        });
    }

    records
}

#[must_use]
pub fn deduplicate_records(records: Vec<NostrDnsRecord>) -> Vec<NostrDnsRecord> {
    let mut seen: std::collections::HashMap<String, NostrDnsRecord> =
        std::collections::HashMap::new();

    for r in records {
        let key = format!("{}:{}:{}:{}", r.fqdn, r.record_type, r.name, r.value);
        match seen.get(&key) {
            Some(existing) if r.created_at <= existing.created_at => {}
            _ => {
                seen.insert(key, r);
            }
        }
    }

    let mut result: Vec<NostrDnsRecord> = seen.into_values().collect();
    result.sort_by_key(|r| std::cmp::Reverse(r.created_at));
    result
}
